sdks/java/io/jdbc/src/main/java/org/apache/beam/sdk/io/jdbc/JdbcIO.java

```diff
@@ -972,11 +979,86 @@ public void populateDisplayData(DisplayData.Builder builder) {
       inner.populateDisplayData(builder);
     }
 
+    private boolean hasStatementAndSetter() {
+      return inner.getStatement() != null && inner.getPreparedStatementSetter() != null;
+    }
+
     @Override
     public PDone expand(PCollection<T> input) {
+      // fixme: validate invalid table input
+      if (input.hasSchema() && !hasStatementAndSetter()) {
+        checkArgument(inner.getTable() != null, "table cannot be null if statement is not provided");
+        Schema schema = input.getSchema();
+        List<Schema.Field> fields = getFilteredFields(schema);
+        inner = inner.withStatement(JdbcUtil.generateStatement(inner.getTable(), fields));
+        inner = inner.withPreparedStatementSetter(new AutoGeneratedPreparedStatementSetter(fields, input.getToRowFunction()));
+      }
+
       inner.expand(input);
       return PDone.in(input.getPipeline());
     }
+
+    private List<Schema.Field> getFilteredFields(Schema schema) {
+      List<Schema.Field> fields = schema.getFields();
+      Connection connection;
+      Schema tableSchema;
+      try {
+        connection = inner.getDataSourceProviderFn().apply(null).getConnection();
+        // Query table to get actual columns
+        ResultSet rs = connection.createStatement()
+                .executeQuery(String.format("SELECT * FROM %s", inner.getTable()));
+        tableSchema = SchemaUtil.toBeamSchema(rs.getMetaData());
+      } catch (Exception e) {
+        throw new RuntimeException("Error while determining columns from table: " + inner.getTable(), e);
+      }
+      fields = fields.stream().filter((field) ->
```

Comment:
If not all the fields from the input source match the output table, there's a potential for silent data loss. I think an exception should be thrown in that case (or, have an option for the user to explicitly acknowledge that some fields will be lost)